### DISCUSSION TITLE
Dual UAC2 Audio devices caused tud_audio_rx_done_pre_read_cb callbacks on audio_n_write

### DISCUSSION THREAD

**Discussion by mkarliner** [hathach/tinyusb#2083] [General]
### Operating System

MacOS

### Board

Raspberry Pi Pico

### Firmware

Based on UAC2 headset example, extended to have 2 UAC2_Headset devices
The headset example works as expected.

### What happened ?

I have two UAC2 devices running.
- A device called 'L' acts as a microphone, sending data from an electret microphone (using the RPi microphone library)
up to the Mac using audio_n_write.
- The Mac runs sox, consuming the mic data and sending it to a wav file. Note the Mac is not outputting to either of the device's speaker channel.
-  The tud_audio_rx_done_pre_read_cb callback is triggered, despite no data being sent to the speaker channels.

### How to reproduce ?

On Mac : sox -r 44100 -t coreaudio "VM Loop" -c 2 test.wav
tud_audio_rx_done_pre_read_cb is triggered.

zipfile attached. evidence.c is a redacted version of main.c for readability.

interface enum and endpoint defs are as follows
num
{
  
  ITF_NUM_AUDIO1_CONTROL = 0,
  ITF_NUM_AUDIO1_STREAMING_SPK,
  ITF_NUM_AUDIO1_STREAMING_MIC,
  //   ITF_NUM_CDC,
  // ITF_CDC_DATA,
  // ITF_NUM_AUDIO2_CONTROL ,
  ITF_NUM_AUDIO2_CONTROL,
  ITF_NUM_AUDIO2_STREAMING_SPK,
  ITF_NUM_AUDIO2_STREAMING_MIC,

  ITF_NUM_TOTAL
};

#define EPNUM_AUDIO1_IN 0x01
#define EPNUM_AUDIO1_OUT 0x01
#define EPNUM_AUDIO2_IN 0x02
#define EPNUM_AUDIO2_OUT 0x02
[evidence.zip](https://github.com/hathach/tinyusb/files/11574661/evidence.zip)

### Debug Log as txt file (LOG/CFG_TUSB_DEBUG=2)

WWW2
  Queue EP 02 with 388 bytes ...
USBD Xfer Complete on EP 82 with 0 bytes
  AUDIO xfer callback
  Queue EP 82 with 0 bytes ...
USBD Xfer Complete on EP 02 with 176 bytes
  AUDIO xfer callback
WWW2
  Queue EP 02 with 388 bytes ...
USBD Xfer Complete on EP 82 with 0 bytes
  AUDIO xfer callback
  Queue EP 82 with 0 bytes ...

USBD Setup Received 01 0B 00 00 04 00 00 00
  Set Interface
  AUDIO control request
  Set itf: 4 - alt: 0
  CLOSING Endpoint: 0x02
Set interface 4 alt 0
  Queue EP 80 with 0 bytes ...
USBD Xfer Complete on EP 82 with 0 bytes
  AUDIO xfer callback
  Queue EP 82 with 0 bytes ...
USBD Xfer Complete on EP 80 with 0 bytes

USBD Setup Received 01 0B 00 00 05 00 00 00
  Set Interface
  AUDIO control request
  Set itf: 5 - alt: 0
  CLOSING Endpoint: 0x82
Set interface 5 alt 0
  Queue EP 80 with 0 bytes ...
USBD Xfer Complete on EP 80 with 0 bytes

### I have checked existing issues, dicussion and documentation

- [X] I confirm I have checked existing issues, dicussion and documentation.

**Comment by HiFiPhile** [comment]
There is no such bug, you must have misconfigured something.

I saw in your `usb_descriptors.c`:
```
// #else
// #define EPNUM_AUDIO1_IN 0x01
// #define EPNUM_AUDIO1_OUT 0x01
// #define EPNUM_AUDIO2_IN 0x07
// #define EPNUM_AUDIO2_OUT 0x07
```
In & out endpoints can't be the same.

**Reply by mkarliner** [reply]
There are a lot of ifdefs I've inhierited from the demo code for various micro's that
have fixed allocation of endpoints. I can't find any specific ones
for the Pi Pico.

`
// #else
// #define EPNUM_AUDIO1_IN 0x01
// #define EPNUM_AUDIO1_OUT 0x01
// #define EPNUM_AUDIO2_IN 0x07
// #define EPNUM_AUDIO2_OUT 0x07
`

Those are commented out, but in the source for the uac2 demo app,
there is 
`#else
  #define EPNUM_AUDIO_IN    0x01
  #define EPNUM_AUDIO_OUT   0x01
#endif`

on line 102. (the in's are or'd with x80 in the macro call) which has worked fine for me.

I use 
`#define EPNUM_AUDIO1_IN 0x01
#define EPNUM_AUDIO1_OUT 0x01
#define EPNUM_AUDIO2_IN 0x02
#define EPNUM_AUDIO2_OUT 0x02`

as listed above. 
Settings EPNUM_AUDIO2_OUT to 0x03 not only doesn't work,
it also freezes my Mac!

In general both devices work in both directions.
I realise that this looks like a dumb user error, but I have done my best to rigorously
check.

I've search in vain for directions on how to allocate endpoint, unit and terminal numbers,
any advice would be greatly appreciated.

**Comment by HiFiPhile** [comment]
At least you should provide buildable source files.

`evidence.c` missing at least includes and interfaces missing in `usb_descriptors`

And many dependency missing.

**Reply by mkarliner** [reply]
Interface enums are in audio_config.h
A revised version of the zip file attached with audio_config.h and main.c (really messy)
evidence.c is a redacted version of my main.c which is such a mess
after weeks of struggle that I thought I'd remove a lot of stuff, evidently
too much :-)

Apologies to all for raising a bug, it does look like misconfiguration.
I'll work some more this week endand drop a cleaned up, but complete version of main.c
Thanks for your patience.
[evidence.zip](https://github.com/hathach/tinyusb/files/11578056/evidence.zip)

**Comment by mkarliner** [comment]
I've done a clean up of the code, and tried yet again to track down the issue, to no avail.
Here is the source directory.
[dual_uac2.zip](https://github.com/hathach/tinyusb/files/11616337/dual_uac2.zip)

**Comment by HiFiPhile** [comment]
Can't reproduce with LPC55S69 on Win10.

Since I don't use Pico nor Mac I can't replicate your setup.

What I suggest is add a log here: https://github.com/hathach/tinyusb/blob/ddc029ce54913fa898c5225a2e87f136a6b81060/src/class/audio/audio_device.c#LL2115C6-L2115C6
```
        TU_LOG1("audiod_rx_done_cb() ep_out:%d ep_in:%d alt:%d\r\n",
          audio->ep_out, audio->ep_in, audio->alt_setting[0]);
```
If you have a debugger to dump `_audiod_fct` when the callback is called is even better.
![image](https://github.com/hathach/tinyusb/assets/4375114/265620ee-2693-484b-9d5c-7f3f1960969b)

**Comment by mkarliner** [comment]
OK, I'm running this command on the Mac,

 sox -r 44100 -t coreaudio "VM Loop" -c 2 test.wav

In other words, just receive from the device.
My main loop reduced to this : 
`
 while(1)
  {
    tud_task();
    tud_audio_n_write(1, (uint8_t *)mic_buf, (uint16_t)1);

    led_blinking_task();
  }

In other words, just keep writing 1 sample .
I've eliminated any reading of the microphone data in case
the DMA causes problems.

`
The log you suggested prints this

`audiod_rx_done_cb() ep_out:6 ep_in:89 alt:1`

BTW, in my copy of tinyusb it's at line 2154

And the dump  of _audiod_fct gives this.

'$2 = {{
    rhport = 0 '\000',
    p_desc = 0x10007c6d <desc_configuration+17> "\t\004",
    ep_in = 0 '\000',
    ep_in_sz = 0,
    ep_in_as_intf_num = 0 '\000',
    ep_out = 0 '\000',
    ep_out_sz = 0,
    ep_out_as_intf_num = 0 '\000',
    desc_length = 312,
    ctrl_buf = 0x20002080 <ctrl_buf_1> "",
    ctrl_buf_sz = 64 '@',
    alt_setting = 0x20000f98 <alt_setting_1> "",
    ep_out_ff = {
      buffer = 0x2000142c <audio_ep_out_sw_buf_1> "",
      depth = 1552,
      {
        item_size = 1,
        overwritable = true
      },
      wr_idx = 0,
      rd_idx = 0,
      mutex_wr = 0x0,
      mutex_rd = 0x20002158 <ep_out_ff_mutex_rd_1>
    },
    ep_in_ff = {
      buffer = 0x20000fa0 <audio_ep_in_sw_buf_1> "",
      depth = 388,
      {
        item_size = 1,
        overwritable = true
      },
      wr_idx = 0,
      rd_idx = 0,
      mutex_wr = 0x20002148 <ep_in_ff_mutex_wr_1>,
      mutex_rd = 0x0
    },
    lin_buf_out = 0x200026f8 <lin_buf_out_1> "",
--Type <RET> for more, q to quit, c to continue without paging--
    lin_buf_in = 0x20002570 <lin_buf_in_1> ""
  }, {
    rhport = 0 '\000',
    p_desc = 0x10007da5 <desc_configuration+329> "\t\004\003",
    ep_in = 137 '\211',
    ep_in_sz = 194,
    ep_in_as_intf_num = 5 '\005',
    ep_out = 6 '\006',
    ep_out_sz = 388,
    ep_out_as_intf_num = 4 '\004',
    desc_length = 312,
    ctrl_buf = 0x200020c0 <ctrl_buf_2> "",
    ctrl_buf_sz = 64 '@',
    alt_setting = 0x20000f9c <alt_setting_2> "\001\001",
    ep_out_ff = {
      buffer = 0x20001a3c <audio_ep_out_sw_buf_2> "",
      depth = 1552,
      {
        item_size = 1,
        overwritable = true
      },
      wr_idx = 0,
      rd_idx = 0,
      mutex_wr = 0x0,
      mutex_rd = 0x20002160 <ep_out_ff_mutex_rd_2>
    },
    ep_in_ff = {
      buffer = 0x20001124 <audio_ep_in_sw_buf_2> "",
      depth = 776,
      {
        item_size = 1,
        overwritable = true
      },
      wr_idx = 234,
      rd_idx = 1388,
      mutex_wr = 0x20002150 <ep_in_ff_mutex_wr_2>,
      mutex_rd = 0x0
    },
    lin_buf_out = 0x20002a00 <lin_buf_out_2> "",
    lin_buf_in = 0x20002634 <lin_buf_in_2> ""
  }}
(gdb) `

So I'm still turning up in the rx_done callback, even though nothing on the Mac is sending stuff back to me.

Hope this helps.

thanks a bunch

Mike

**Comment by HiFiPhile** [comment]
From your log I see no corruption in the USBD layer, UAC driver was invoked like a real packed arrived.

There could be something wrong in the RP2040 dcd driver.

Reopened the issue for RP2040.
